src/logging/logging.ts

```diff
@@ -65,128 +27,61 @@ module Logging {
-        // trim from the head 10 percent each time.
-        logBuffer.splice(0, MAX_BUFFERED_LOG / 10);
-      }
-      logBuffer.push(Message);
+    if (!(tag in waiters)) {
+      waiters[tag] = [];
     }
-  }
-
-  // Gets log as a encrypted blob, which can be transported in insecure
-  // channel.
-  export function getEncrypedLogBuffer(tags:string[]) : ArrayBuffer {
-    // TODO: to be implemented.
-    return new ArrayBuffer(0);
-  }
-
-  // Gets log in plaintext, which should really be used in development env
-  // only.
-  // Usage: getLogs(['network', 'xmpp']);
-  // It will return log message with tag 'netowrk' and 'xmpp' only.
-  // getLogs() without specify any tag will return all messages.
-  export function getLogs(tags?:string[]) : string[] {
-    // TODO: use input to select log message.
-    if(!tags || tags.length === 0) {
-      return logBuffer.map(formatMessage);
-    } else {
-      return logBuffer.filter((m:Message) => {
-        return tags.indexOf(m.tag) >= 0;
-      }).map(formatMessage);
-    }
-  }
-
-  // Clears all the logs stored in buffer.
-  export function clearLogs() : void {
-    logBuffer = [];
-  }
-  // Enables/Disables log facility.
-  export function enable() : void {
-    enabled = true;
-  }
-  // Enables/Disables log facility.
-  export function disable() : void {
-    enabled = false;
-  }
-
-  // Sets the log filter for console output. Caller can specify logs of
-  // desired tags and levels for console output.
-  // Usage example: setConsoleFilter("*:E", "network:D")
-  // It means: output message in Error level for any module
-  //           output message in debug level and above for "network" module.
-  export function setConsoleFilter(args: string[]) : void {
-    consoleFilter = {};
-    for (var i = 0; i < args.length; i++) {
-      var parts = args[i].split(':');
-      consoleFilter[parts[0]] = parts[1];
-    }
-  }
-
-  // Sets the log filter for buffered log.
-  // Usage example: setBufferedLogFilter("*:E", "network:D")
-  // It means: buffer message in Error level for any module
-  //           buffer message in debug level and above for "network" module.
-  export function setBufferedLogFilter(args: string[]) : void {
-    bufferedLogFilter = {};
-    for (var i = 0; i < args.length; i++) {
-      var parts = args[i].split(':');
-      bufferedLogFilter[parts[0]] = parts[1];
+    waiters[tag].push({level: level, msg: formatStringMessageWithArgs_(msg, args)});
+  }
+  
+  function registerLogger(tag:string, logger:freedom.Logger) : void {
+    loggers[tag] = logger;
+    if (tag in waiters) {
+      for (var i = 0; i < waiters[tag].length; i += 1) {
+        var msg:Msg = waiters[tag][i];
+        doLog(msg.level, tag, msg.msg);
+      }
+      delete waiters[tag];
     }
   }
 
   export class Log {
-    constructor(private tag_:string) {}
+    constructor(private tag_:string) {
+      freedom['core']().getLogger(this.tag_).then((logger:freedom.Logger) => {
```

Comment:
A member variable of type Promise.

It's arguable how understandable the `waiters` array is but it's definitely more complex than introducing a member variable to `Log`.